In [1]:
import numpy as np
import math
from models import interpolation, SRCNN_train, SRCNN_model, SRCNN_predict, DNCNN_train, DNCNN_model, DNCNN_predict
#from scipy.misc import imresize
from scipy.io import loadmat
import matplotlib.pyplot as plt

Using TensorFlow backend.


In [2]:
import os
os.path.isfile("../dataset/Perfect_H_40000.mat")  #
os.path.isfile("../dataset/My_noisy_H_22.mat")

True

In [3]:
# data:Perfect_H_40000.mat
# noisy: My_noisy_H_22.mat
if __name__ == "__main__":
    # load datasets 
    channel_model = "VehA"
    SNR = 22
    Number_of_pilots = 48
    perfect = loadmat("../dataset/Perfect_H_40000.mat")['My_perfect_H']
    noisy_input = loadmat("../dataset/My_noisy_H_22.mat")['My_noisy_H']

In [4]:
     interp_noisy = interpolation(noisy_input , SNR , Number_of_pilots , 'rbf')


In [5]:
#     interp_noisy = np.load("interp_noisy.npy")
#     np.save('interp_noisy.npy',interp_noisy)

In [7]:
perfect_image = np.zeros((len(perfect),72,14,2))
perfect_image[:,:,:,0] = np.real(perfect)
perfect_image[:,:,:,1] = np.imag(perfect)
perfect_image = np.concatenate((perfect_image[:,:,:,0], perfect_image[:,:,:,1]), axis=0).reshape(2*len(perfect), 72, 14, 1)

In [8]:
####### ------ training SRCNN ------ #######

idx_random = np.random.rand(len(perfect_image)) < (1/9)  # uses 32000 from 36000 as training and the rest as validation
train_data, train_label = interp_noisy[idx_random,:,:,:] , perfect_image[idx_random,:,:,:]
val_data, val_label = interp_noisy[~idx_random,:,:,:] , perfect_image[~idx_random,:,:,:]

In [9]:
import tensorflow as tf
print("Num GPUs Available: ", len(tf.config.experimental.list_physical_devices('GPU')))

Num GPUs Available:  1


In [15]:
# SRCNN_train(train_data ,train_label, val_data , val_label , channel_model , Number_of_pilots , SNR )

Instructions for updating:
If using Keras pass *_constraint arguments to layers.


/home/file/source/Channel-Estimation_test/models.py:98: UserWarning: Update your `Conv2D` call to the Keras 2 API: `Conv2D(64, (9, 9), activation="relu", padding="same", kernel_initializer="he_normal")`
  c2 = Convolution2D( 32 , 1 , 1 , activation = 'relu', init = 'he_normal', border_mode='same')(c1)
/home/file/source/Channel-Estimation_test/models.py:99: UserWarning: Update your `Conv2D` call to the Keras 2 API: `Conv2D(32, (1, 1), activation="relu", padding="same", kernel_initializer="he_normal")`
  c3 = Convolution2D( 1 , 5 , 5 , init = 'he_normal', border_mode='same')(c2)
/home/file/source/Channel-Estimation_test/models.py:100: UserWarning: Update your `Conv2D` call to the Keras 2 API: `Conv2D(1, (5, 5), padding="same", kernel_initializer="he_normal")`
  #c4 = Input(shape = input_shape)(c3)
/home/file/source/Channel-Estimation_test/models.py:102: UserWarning: Update your `Model` call to the Keras 2 API: `Model(inputs=Tensor("in..., outputs=Tensor("co...)`
  ##compile


Model: "model_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_1 (InputLayer)         (None, 72, 14, 1)         0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 72, 14, 64)        5248      
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 72, 14, 32)        2080      
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 72, 14, 1)         801       
Total params: 8,129
Trainable params: 8,129
Non-trainable params: 0
_________________________________________________________________
None


Epoch 00001: val_loss improved from inf to 0.01143, saving model to SRCNN_check.h5

Epoch 00002: val_loss improved from 0.01143 to 0.00551, saving model to SRCNN_check.h5

Epoch 00003: val_loss improved from 0.00551 to 0.00436, saving model to SRCN


Epoch 00112: val_loss did not improve from 0.00147

Epoch 00113: val_loss did not improve from 0.00147

Epoch 00114: val_loss did not improve from 0.00147

Epoch 00115: val_loss improved from 0.00147 to 0.00145, saving model to SRCNN_check.h5

Epoch 00116: val_loss did not improve from 0.00145

Epoch 00117: val_loss did not improve from 0.00145

Epoch 00118: val_loss did not improve from 0.00145

Epoch 00119: val_loss did not improve from 0.00145

Epoch 00120: val_loss did not improve from 0.00145

Epoch 00121: val_loss did not improve from 0.00145

Epoch 00122: val_loss did not improve from 0.00145

Epoch 00123: val_loss did not improve from 0.00145

Epoch 00124: val_loss did not improve from 0.00145

Epoch 00125: val_loss did not improve from 0.00145

Epoch 00126: val_loss did not improve from 0.00145

Epoch 00127: val_loss did not improve from 0.00145

Epoch 00128: val_loss did not improve from 0.00145

Epoch 00129: val_loss did not improve from 0.00145

Epoch 00130: val_loss impro


Epoch 00263: val_loss did not improve from 0.00133

Epoch 00264: val_loss did not improve from 0.00133

Epoch 00265: val_loss did not improve from 0.00133

Epoch 00266: val_loss did not improve from 0.00133

Epoch 00267: val_loss did not improve from 0.00133

Epoch 00268: val_loss did not improve from 0.00133

Epoch 00269: val_loss did not improve from 0.00133

Epoch 00270: val_loss did not improve from 0.00133

Epoch 00271: val_loss did not improve from 0.00133

Epoch 00272: val_loss did not improve from 0.00133

Epoch 00273: val_loss did not improve from 0.00133

Epoch 00274: val_loss did not improve from 0.00133

Epoch 00275: val_loss did not improve from 0.00133

Epoch 00276: val_loss did not improve from 0.00133

Epoch 00277: val_loss did not improve from 0.00133

Epoch 00278: val_loss did not improve from 0.00133

Epoch 00279: val_loss did not improve from 0.00133

Epoch 00280: val_loss did not improve from 0.00133

Epoch 00281: val_loss did not improve from 0.00133

Epoch 00282

After training 300 epochs, the val_loss improved from inf to 0.00133, saving model to SRCNN_VehA_48_22.h5.

In [10]:
####### ------ prediction using SRCNN ------ #######
srcnn_pred_train = SRCNN_predict(train_data, channel_model , Number_of_pilots , SNR) 

Instructions for updating:
If using Keras pass *_constraint arguments to layers.


/home/file/source/Channel-Estimation_test/models.py:97: UserWarning: Update your `Conv2D` call to the Keras 2 API: `Conv2D(64, (9, 9), activation="relu", padding="same", kernel_initializer="he_normal")`
  c1 = Convolution2D( 64 , 9 , 9 , activation = 'relu', init = 'he_normal', border_mode='same')(x)
/home/file/source/Channel-Estimation_test/models.py:98: UserWarning: Update your `Conv2D` call to the Keras 2 API: `Conv2D(32, (1, 1), activation="relu", padding="same", kernel_initializer="he_normal")`
  c2 = Convolution2D( 32 , 1 , 1 , activation = 'relu', init = 'he_normal', border_mode='same')(c1)
/home/file/source/Channel-Estimation_test/models.py:99: UserWarning: Update your `Conv2D` call to the Keras 2 API: `Conv2D(1, (5, 5), padding="same", kernel_initializer="he_normal")`
  c3 = Convolution2D( 1 , 5 , 5 , init = 'he_normal', border_mode='same')(c2)
/home/file/source/Channel-Estimation_test/models.py:101: UserWarning: Update your `Model` call to the Keras 2 API: `Model(inputs=Tenso

In [11]:
 srcnn_pred_validation = SRCNN_predict(val_data, channel_model , Number_of_pilots , SNR)

In [13]:
print(srcnn_pred_train.shape)

(8872, 72, 14, 1)


In [14]:
# save the result of prediction
np.savetxt('srcnn_pred_train.txt',srcnn_pred_train[0][0])
np.savetxt('srcnn_pred_validation.txt',srcnn_pred_validation[0][0])

In [15]:
####### ------ training DNCNN ------ #######
DNCNN_train(train_data ,train_label, val_data , val_label , channel_model , Number_of_pilots , SNR )

Model: "model_3"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
input_3 (InputLayer)            (None, None, None, 1 0                                            
__________________________________________________________________________________________________
conv2d_7 (Conv2D)               (None, None, None, 6 640         input_3[0][0]                    
__________________________________________________________________________________________________
activation_1 (Activation)       (None, None, None, 6 0           conv2d_7[0][0]                   
__________________________________________________________________________________________________
conv2d_8 (Conv2D)               (None, None, None, 6 36928       activation_1[0][0]               
____________________________________________________________________________________________


Epoch 00001: val_loss improved from inf to 0.00935, saving model to DNCNN_check.h5

Epoch 00002: val_loss improved from 0.00935 to 0.00865, saving model to DNCNN_check.h5

Epoch 00003: val_loss improved from 0.00865 to 0.00743, saving model to DNCNN_check.h5

Epoch 00004: val_loss improved from 0.00743 to 0.00638, saving model to DNCNN_check.h5

Epoch 00005: val_loss improved from 0.00638 to 0.00494, saving model to DNCNN_check.h5

Epoch 00006: val_loss improved from 0.00494 to 0.00368, saving model to DNCNN_check.h5

Epoch 00007: val_loss improved from 0.00368 to 0.00336, saving model to DNCNN_check.h5

Epoch 00008: val_loss improved from 0.00336 to 0.00220, saving model to DNCNN_check.h5

Epoch 00009: val_loss improved from 0.00220 to 0.00194, saving model to DNCNN_check.h5

Epoch 00010: val_loss improved from 0.00194 to 0.00177, saving model to DNCNN_check.h5

Epoch 00011: val_loss improved from 0.00177 to 0.00164, saving model to DNCNN_check.h5

Epoch 00012: val_loss did not impro


Epoch 00143: val_loss did not improve from 0.00106

Epoch 00144: val_loss did not improve from 0.00106

Epoch 00145: val_loss did not improve from 0.00106

Epoch 00146: val_loss did not improve from 0.00106

Epoch 00147: val_loss did not improve from 0.00106

Epoch 00148: val_loss did not improve from 0.00106

Epoch 00149: val_loss did not improve from 0.00106

Epoch 00150: val_loss did not improve from 0.00106

Epoch 00151: val_loss did not improve from 0.00106

Epoch 00152: val_loss did not improve from 0.00106

Epoch 00153: val_loss did not improve from 0.00106

Epoch 00154: val_loss did not improve from 0.00106

Epoch 00155: val_loss did not improve from 0.00106

Epoch 00156: val_loss did not improve from 0.00106

Epoch 00157: val_loss did not improve from 0.00106

Epoch 00158: val_loss did not improve from 0.00106

Epoch 00159: val_loss did not improve from 0.00106

Epoch 00160: val_loss did not improve from 0.00106

Epoch 00161: val_loss did not improve from 0.00106

Epoch 00162

After training 200 epochs, the val_loss improved from 0.0935 to 0.00106, saving model to DNCNN_VehA_48_22.h5.

In [ ]:
`